File: crates/api_clients/artcraft/artcraft_api_defs/src/generate/cost_estimate/estimate_video_cost.rs
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use enums::common::generation::common_aspect_ratio::CommonAspectRatio;
use enums::common::generation::common_video_model::CommonVideoModel;
use enums::common::generation::common_resolution::CommonResolution;
use enums::common::generation_provider::GenerationProvider;

pub const ESTIMATE_VIDEO_COST_PATH: &str = "/v1/generate/cost_estimate/video";

/// Request body for the video cost estimate endpoint.
///
/// Every model cares about its own set of parameters and not all
/// parameters may be relevant for every model.
#[derive(Serialize, Deserialize, ToSchema, Debug)]
pub struct EstimateVideoCostRequest {
  /// The video model to estimate costs for.
  pub model: CommonVideoModel,

  /// The provider to route the generation through.
  pub provider: GenerationProvider,

  /// The type of generation (determines how many input images are involved).
  /// This is a tagged enum, so it looks like:
  ///   "generation_mode": {"type": "text_to_video"}
  ///   "generation_mode": {"type": "reference_image_to_video", "count": 1}
  pub generation_mode: GenerationMode,

  /// Optional aspect ratio.
  pub aspect_ratio: Option<CommonAspectRatio>,

  /// Optional resolution.
  pub resolution: Option<CommonResolution>,

  /// Duration in seconds.
  pub duration_seconds: Option<u16>,

  /// Number of videos to generate in parallel.
  pub video_batch_count: Option<u16>,

  /// Whether to generate audio for the video.
  pub generate_audio: Option<bool>,
}

/// Describes the type of video generation being requested.
#[derive(Serialize, Deserialize, ToSchema, Debug)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum GenerationMode {
  TextToVideo,
  StartFrameToVideo,
  StartAndEndFrameToVideo,
  ReferenceImageToVideo { count: u32 },
}

/// Response body for the video cost estimate endpoint.
#[derive(Serialize, Deserialize, ToSchema)]
pub struct EstimateVideoCostResponse {
  pub success: bool,

  /// Estimated cost in credits.
  pub cost_in_credits: Option<u64>,

  /// Estimated cost in USD cents.
  pub cost_in_usd_cents: Option<u64>,

  /// Whether the generation is free for this user/plan.
  pub is_free: bool,

  /// Whether the user has unlimited generations.
  pub is_unlimited: bool,

  /// Whether the user is rate limited.
  pub is_rate_limited: bool,

  /// Whether the output will have a watermark.
  pub has_watermark: bool,
}

/// Error response for the video cost estimate endpoint.
#[derive(Serialize, Deserialize, ToSchema)]
pub struct EstimateVideoCostError {
  pub success: bool,
  pub error_type: EstimateVideoCostErrorType,
  pub error_message: String,
}

#[derive(Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum EstimateVideoCostErrorType {
  InvalidProviderForModel,
  InvalidInput,
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_serialization_tagging1() {
    let request = EstimateVideoCostRequest {
      model: CommonVideoModel::GrokVideo,
      provider: GenerationProvider::Artcraft,
      generation_mode: GenerationMode::TextToVideo,
      aspect_ratio: None,
      resolution: None,
      duration_seconds: None,
      video_batch_count: None,
      generate_audio: None,
    };
    let serialized = serde_json::to_string(&request).unwrap();
    assert!(serialized.contains("\"generation_mode\":{\"type\":\"text_to_video\"}"));
    //assert_eq!(serialized, "{}");
  }

  #[test]
  fn test_serialization_tagging2() {
    let request = EstimateVideoCostRequest {
      model: CommonVideoModel::GrokVideo,
      provider: GenerationProvider::Artcraft,
      generation_mode: GenerationMode::ReferenceImageToVideo { count: 1 },
      aspect_ratio: None,
      resolution: None,
      duration_seconds: None,
      video_batch_count: None,
      generate_audio: None,
    };
    let serialized = serde_json::to_string(&request).unwrap();
    assert!(serialized.contains("\"generation_mode\":{\"type\":\"reference_image_to_video\",\"count\":1}"));
  }
}

File: frontend/libs/tauri-api/src/lib/cost_estimate/EstimateVideoCost.ts
import { invoke } from "@tauri-apps/api/core";
import { CommandResult, CommandSuccessStatus } from "../common/CommandStatus";
import {
  CommonVideoModel,
  CommonAspectRatio,
  CommonVideoResolution,
  GenerationMode,
  GenerationProvider,
} from "@storyteller/api-enums";

export interface EstimateVideoCostRequest {
  model: CommonVideoModel;
  provider: GenerationProvider;
  generation_mode: GenerationMode;
  aspect_ratio?: CommonAspectRatio;
  resolution?: CommonVideoResolution;
  duration_seconds?: number;
  video_batch_count?: number;
  generate_audio?: boolean;
}

export interface EstimateVideoCostPayload {
  success: boolean;
  cost_in_credits?: number;
  cost_in_usd_cents?: number;
  is_free: boolean;
  is_unlimited: boolean;
  is_rate_limited: boolean;
  has_watermark: boolean;
}

export interface EstimateVideoCostSuccess extends CommandResult {
  payload: EstimateVideoCostPayload;
}

export interface EstimateVideoCostErrorPayload {
  success: boolean;
  error_type: "invalid_provider_for_model" | "invalid_input";
  error_message: string;
}

export interface EstimateVideoCostErrorResult extends CommandResult {
  error_details?: EstimateVideoCostErrorPayload;
}

export type EstimateVideoCostResult =
  | EstimateVideoCostSuccess
  | EstimateVideoCostErrorResult;

export const EstimateVideoCost = async (
  request: EstimateVideoCostRequest,
): Promise<EstimateVideoCostResult> => {
  const result = await invoke("estimate_video_cost_command", { request });
  return result as EstimateVideoCostResult;
};

export function isEstimateVideoCostSuccess(
  r: EstimateVideoCostResult,
): r is EstimateVideoCostSuccess {
  return r.status === CommandSuccessStatus.Success;
}
